File: TDE2-python/lp_problem.py
"""lp_problem.py -- the LP model plus the two input paths (file and console).

This is just the "model": objective sense, the objective coefficients and the
constraint rows. All SIMPLEX machinery lives in simplex.py so this file stays
easy to read and explain.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import List


class Objective(Enum):
    MAX = "max"
    MIN = "min"


class ConstraintType(Enum):
    LE = "<="
    GE = ">="
    EQ = "="


@dataclass
class LPProblem:
    """An LP in original form:

        max / min   Z = c[0]*x1 + ... + c[n-1]*xn
        subject to  A[i] . x  (rel[i])  b[i]      for i = 0..m-1
                    x >= 0
    """
    objective: Objective = Objective.MAX
    n: int = 0                                       # decision variables
    m: int = 0                                       # constraints
    c: List[float] = field(default_factory=list)     # objective coefficients (n)
    A: List[List[float]] = field(default_factory=list)  # constraint matrix (m x n)
    rel: List[ConstraintType] = field(default_factory=list)  # relations (m)
    b: List[float] = field(default_factory=list)     # right-hand sides (m)


class ParseError(Exception):
    """Raised when the input cannot be understood."""


# Forgiving spellings, like the C++ version.
_REL = {"<=": ConstraintType.LE, "<": ConstraintType.LE,
        ">=": ConstraintType.GE, ">": ConstraintType.GE,
        "=": ConstraintType.EQ, "==": ConstraintType.EQ}

_OBJ = {"max": Objective.MAX, "maximize": Objective.MAX,
        "min": Objective.MIN, "minimize": Objective.MIN}


def _tokenize(text: str) -> List[str]:
    """Flatten text into whitespace-separated tokens, dropping '#' comments."""
    toks: List[str] = []
    for line in text.splitlines():
        hash_i = line.find("#")
        if hash_i != -1:
            line = line[:hash_i]
        toks.extend(line.split())
    return toks


def load_problem_from_file(path: str) -> LPProblem:
    """Read a problem from a text file (format documented in TUTORIAL.txt).

    Token order:  max|min, then n m, then c1..cn, then for each constraint
    a1..an  <rel>  b.
    """
    try:
        with open(path, "r") as f:
            toks = _tokenize(f.read())
    except OSError:
        raise ParseError(f"nao foi possivel abrir o arquivo: {path}")

    pos = 0

    def nxt(what: str) -> str:
        nonlocal pos
        if pos >= len(toks):
            raise ParseError(f"entrada incompleta: esperava {what}")
        t = toks[pos]
        pos += 1
        return t

    def num(what: str) -> float:
        t = nxt(what)
        try:
            return float(t)
        except ValueError:
            raise ParseError(f"valor numerico invalido para {what}: '{t}'")

    p = LPProblem()

    o = nxt("'max' ou 'min'").lower()
    if o not in _OBJ:
        raise ParseError(f"primeiro token deve ser 'max' ou 'min', veio '{o}'")
    p.objective = _OBJ[o]

    try:
        p.n = int(nxt("n"))
        p.m = int(nxt("m"))
    except ValueError:
        raise ParseError("n e m devem ser inteiros")
    if p.n <= 0 or p.m <= 0:
        raise ParseError("n e m devem ser positivos")

    p.c = [num(f"c{j + 1}") for j in range(p.n)]

    for r in range(p.m):
        row = [num(f"a[{r + 1}][{j + 1}]") for j in range(p.n)]
        rel_tok = nxt(f"relacao da restricao {r + 1}")
        if rel_tok not in _REL:
            raise ParseError(f"relacao invalida '{rel_tok}' na restricao {r + 1}")
        p.A.append(row)
        p.rel.append(_REL[rel_tok])
        p.b.append(num(f"b{r + 1}"))

    validate_problem(p)
    return p


def _read_int(prompt: str, positive: bool = False) -> int:
    while True:
        s = input(prompt).strip()
        try:
            v = int(s)
        except ValueError:
            print("  digite um numero inteiro.")
            continue
        if positive and v <= 0:
            print("  digite um inteiro positivo.")
            continue
        return v


def read_problem_from_console() -> LPProblem:
    """Read a problem interactively from stdin (guided prompts).

    Each numeric line is split on whitespace, so the user may type the whole
    objective (e.g. "3 5") or a whole constraint (e.g. "1 0 <= 4") on one line.
    """
    p = LPProblem()

    o = input("Objetivo (max/min): ").strip().lower()
    while o not in _OBJ:
        o = input("  digite 'max' ou 'min': ").strip().lower()
    p.objective = _OBJ[o]

    p.n = _read_int("Numero de variaveis de decisao (n): ", positive=True)
    p.m = _read_int("Numero de restricoes (m): ", positive=True)

    # Objective coefficients, all on one line.
    while True:
        toks = input(f"Coeficientes do objetivo (c1..c{p.n}, separados por espaco): ").split()
        if len(toks) != p.n:
            print(f"  esperados {p.n} valores, recebidos {len(toks)}.")
            continue
        try:
            p.c = [float(t) for t in toks]
            break
        except ValueError:
            print("  valores numericos invalidos, tente de novo.")

    # Each constraint on one line:  a1 .. an  REL  b
    for r in range(p.m):
        while True:
            toks = input(f"Restricao {r + 1} (a1..a{p.n}  REL(<=,>=,=)  b): ").split()
            if len(toks) != p.n + 2:
                print(f"  esperados {p.n + 2} campos (n coeficientes + relacao + b), "
                      f"recebidos {len(toks)}.")
                continue
            rel_tok = toks[p.n]
            if rel_tok not in _REL:
                print(f"  relacao invalida '{rel_tok}', use <=, >= ou =.")
                continue
            try:
                row = [float(t) for t in toks[:p.n]]
                bval = float(toks[p.n + 1])
            except ValueError:
                print("  valores numericos invalidos, tente de novo.")
                continue
            p.A.append(row)
            p.rel.append(_REL[rel_tok])
            p.b.append(bval)
            break

    validate_problem(p)
    return p


def validate_problem(p: LPProblem) -> None:
    """Basic dimension checks. Raises ParseError if something is inconsistent."""
    if p.n <= 0 or p.m <= 0:
        raise ParseError("n e m devem ser positivos")
    if len(p.c) != p.n:
        raise ParseError("numero de coeficientes do objetivo != n")
    if len(p.A) != p.m or len(p.rel) != p.m or len(p.b) != p.m:
        raise ParseError("numero de restricoes inconsistente")
    for r, row in enumerate(p.A):
        if len(row) != p.n:
            raise ParseError(f"restricao {r + 1} tem numero de coeficientes errado")

File: TDE2/src/lp_problem.cpp
// lp_problem.cpp  -- problem model + the two input paths (file and console).
#include "lp_problem.h"

#include <fstream>
#include <iostream>
#include <sstream>

const char* constraintSymbol(ConstraintType t) {
    switch (t) {
        case ConstraintType::LE: return "<=";
        case ConstraintType::GE: return ">=";
        case ConstraintType::EQ: return "=";
    }
    return "?";
}

// Turn a relation token from the input into a ConstraintType.
// Accepts the common spellings so the file format is forgiving.
static bool parseRelation(const std::string& tok, ConstraintType& out) {
    if (tok == "<=" || tok == "<") { out = ConstraintType::LE; return true; }
    if (tok == ">=" || tok == ">") { out = ConstraintType::GE; return true; }
    if (tok == "="  || tok == "==") { out = ConstraintType::EQ; return true; }
    return false;
}

static bool parseObjective(const std::string& tok, Objective& out) {
    if (tok == "max" || tok == "MAX" || tok == "maximize") { out = Objective::MAX; return true; }
    if (tok == "min" || tok == "MIN" || tok == "minimize") { out = Objective::MIN; return true; }
    return false;
}

// ---------------------------------------------------------------------------
// File input.
//
// We first flatten the file into a stream of tokens, dropping blank lines and
// anything after a '#' comment marker. Then we consume tokens in order:
//
//   <max|min>
//   <n> <m>
//   c1 c2 ... cn
//   (for each of m constraints)  a1 a2 ... an  <rel>  b
// ---------------------------------------------------------------------------
bool loadProblemFromFile(const std::string& path, LPProblem& out, std::string& err) {
    std::ifstream in(path);
    if (!in) { err = "could not open file: " + path; return false; }

    std::vector<std::string> tok;
    std::string line;
    while (std::getline(in, line)) {
        // Strip comments: everything from '#' to end of line.
        std::size_t hash = line.find('#');
        if (hash != std::string::npos) line = line.substr(0, hash);
        std::istringstream ls(line);
        std::string t;
        while (ls >> t) tok.push_back(t);
    }

    std::size_t i = 0;
    auto next = [&](std::string& dst) -> bool {
        if (i >= tok.size()) return false;
        dst = tok[i++];
        return true;
    };

    std::string s;
    if (!next(s) || !parseObjective(s, out.objective)) {
        err = "expected 'max' or 'min' as the first token"; return false;
    }

    std::string sn, sm;
    if (!next(sn) || !next(sm)) { err = "expected 'n m' (variable and constraint counts)"; return false; }
    try {
        out.n = std::stoi(sn);
        out.m = std::stoi(sm);
    } catch (...) { err = "n and m must be integers"; return false; }
    if (out.n <= 0 || out.m <= 0) { err = "n and m must be positive"; return false; }

    out.c.resize(out.n);
    for (int j = 0; j < out.n; ++j) {
        if (!next(s)) { err = "not enough objective coefficients"; return false; }
        try { out.c[j] = std::stod(s); } catch (...) { err = "bad objective coefficient: " + s; return false; }
    }

    out.A.assign(out.m, std::vector<double>(out.n, 0.0));
    out.rel.resize(out.m);
    out.b.resize(out.m);
    for (int r = 0; r < out.m; ++r) {
        for (int j = 0; j < out.n; ++j) {
            if (!next(s)) { err = "not enough coefficients for constraint " + std::to_string(r + 1); return false; }
            try { out.A[r][j] = std::stod(s); } catch (...) { err = "bad coefficient: " + s; return false; }
        }
        if (!next(s) || !parseRelation(s, out.rel[r])) {
            err = "expected a relation (<=, >=, =) in constraint " + std::to_string(r + 1); return false;
        }
        if (!next(s)) { err = "missing right-hand side in constraint " + std::to_string(r + 1); return false; }
        try { out.b[r] = std::stod(s); } catch (...) { err = "bad right-hand side: " + s; return false; }
    }

    return validateProblem(out, err);
}

// ---------------------------------------------------------------------------
// Console input (guided). Used when the program is launched with no file.
// ---------------------------------------------------------------------------
LPProblem readProblemFromConsole() {
    LPProblem p;
    std::string s;

    std::cout << "Objetivo (max/min): ";
    std::cin >> s;
    Objective obj;
    while (!parseObjective(s, obj)) {
        std::cout << "  digite 'max' ou 'min': ";
        std::cin >> s;
    }
    p.objective = obj;

    std::cout << "Numero de variaveis de decisao (n): ";
    std::cin >> p.n;
    std::cout << "Numero de restricoes (m): ";
    std::cin >> p.m;

    p.c.resize(p.n);
    std::cout << "Coeficientes da funcao objetivo (c1..c" << p.n << "):\n";
    for (int j = 0; j < p.n; ++j) {
        std::cout << "  c" << (j + 1) << " = ";
        std::cin >> p.c[j];
    }

    p.A.assign(p.m, std::vector<double>(p.n, 0.0));
    p.rel.resize(p.m);
    p.b.resize(p.m);
    for (int r = 0; r < p.m; ++r) {
        std::cout << "Restricao " << (r + 1) << ":\n";
        for (int j = 0; j < p.n; ++j) {
            std::cout << "  a" << (r + 1) << (j + 1) << " = ";
            std::cin >> p.A[r][j];
        }
        std::cout << "  relacao (<=, >=, =): ";
        std::cin >> s;
        ConstraintType ct;
        while (!parseRelation(s, ct)) {
            std::cout << "    use <=, >= ou = : ";
            std::cin >> s;
        }
        p.rel[r] = ct;
        std::cout << "  b" << (r + 1) << " = ";
        std::cin >> p.b[r];
    }
    return p;
}

bool validateProblem(const LPProblem& p, std::string& err) {
    if (p.n <= 0 || p.m <= 0) { err = "n and m must be positive"; return false; }
    if ((int)p.c.size() != p.n) { err = "objective coefficient count != n"; return false; }
    if ((int)p.A.size() != p.m) { err = "constraint row count != m"; return false; }
    if ((int)p.rel.size() != p.m || (int)p.b.size() != p.m) { err = "relation/RHS count != m"; return false; }
    for (int r = 0; r < p.m; ++r) {
        if ((int)p.A[r].size() != p.n) {
            err = "constraint " + std::to_string(r + 1) + " has wrong coefficient count"; return false;
        }
    }
    return true;
}
